src/World.cpp

```diff
@@ -572,50 +572,147 @@ void cWorld::Start(void)
+		if ((Biome == EMCSBiome::biOcean) || (Biome == EMCSBiome::biFrozenOcean))
+		{
+			BiomeOffset += Vector3d(cChunkDef::Width * 4, 0, 0);
+			continue;
+		}
 
-		if (
-			(biome != biOcean) && (biome != biFrozenOcean) &&  // The biome is acceptable (don't want a small ocean island)
-			!IsBlockWaterOrIce(GetBlock(SpawnX, GetHeight(SpawnX, SpawnZ), SpawnZ))  // The terrain is acceptable (don't want to spawn inside a lake / river)
-		)
+		// Found a usable biome
+		// Spawn chunks so we can find a nice spawn.
+		int ChunkX = 0, ChunkZ = 0;
+		cChunkDef::BlockToChunk(BiomeOffset.x, BiomeOffset.z, ChunkX, ChunkZ);
+		cSpawnPrepare::PrepareChunks(*this, ChunkX, ChunkZ, a_MaxSpawnRadius);
+		break;
+	} while ((--BiomeCheckCount) > 0);
+	
+	// Check 0, 0 first.
+	double SpawnY = 0.0;
+	if (CanSpawnAt(BiomeOffset.x, BiomeOffset.z, SpawnY))
+	{
+		m_SpawnX = BiomeOffset.x + 0.5;
+		m_SpawnY = SpawnY;
+		m_SpawnZ = BiomeOffset.z + 0.5;
+
+		LOGINFO("Generated spawnpoint position at {%i, %i, %i}", static_cast<int>(m_SpawnX), static_cast<int>(m_SpawnY), static_cast<int>(m_SpawnZ));
+		return;
+	}
+
+	// A search grid (searches clockwise around the origin)
+	static const Vector3i ChunkOffset[] =
+	{
+		Vector3i(0, 0, cChunkDef::Width),
+		Vector3i(cChunkDef::Width, 0, cChunkDef::Width),
+		Vector3i(cChunkDef::Width, 0, 0),
+		Vector3i(cChunkDef::Width, 0, -cChunkDef::Width),
+		Vector3i(0, 0, -cChunkDef::Width),
+		Vector3i(-cChunkDef::Width, 0, -cChunkDef::Width),
+		Vector3i(-cChunkDef::Width, 0, 0),
+		Vector3i(-cChunkDef::Width, 0, cChunkDef::Width),
+	};
+
+	static const int PerRadiSearchCount = ARRAYCOUNT(ChunkOffset);
+
+	for (int RadiusOffset = 1; RadiusOffset < a_MaxSpawnRadius; ++RadiusOffset)
+	{
+		for (int SearchGridIndex = 0; SearchGridIndex < PerRadiSearchCount; ++SearchGridIndex)
+		{
+			const Vector3i PotentialSpawn = BiomeOffset + (ChunkOffset[SearchGridIndex] * RadiusOffset);
+			
+			if (CanSpawnAt(PotentialSpawn.x, PotentialSpawn.z, SpawnY))
+			{
+				m_SpawnX = PotentialSpawn.x + 0.5;
+				m_SpawnY = SpawnY;
+				m_SpawnZ = PotentialSpawn.z + 0.5;
+
+				LOGINFO("Generated spawnpoint position at {%i, %i, %i}", static_cast<int>(m_SpawnX), static_cast<int>(m_SpawnY), static_cast<int>(m_SpawnZ));
+				return;
+			}
+		}
+	}
+
+	m_SpawnY = GetHeight(static_cast<int>(m_SpawnX), static_cast<int>(m_SpawnZ));
+	LOGWARNING("Did not find an acceptable spawnpoint. Generated a random spawnpoint position at {%i, %i, %i}", static_cast<int>(m_SpawnX), static_cast<int>(m_SpawnY), static_cast<int>(m_SpawnZ));
+}
+
+
+
+
+
+bool cWorld::CanSpawnAt(double a_X, double a_Z, double &a_Y)
+{
+	static const BLOCKTYPE ValidSpawnBlocks[] =
+	{
+		E_BLOCK_GRASS,
+		E_BLOCK_DIRT,
+		E_BLOCK_SAND,
+		E_BLOCK_SNOW,
+		E_BLOCK_SNOW_BLOCK,
+		E_BLOCK_NETHERRACK
+	};
+
+	static const int ValidSpawnBlocksCount = ARRAYCOUNT(ValidSpawnBlocks);
+
+	for (int PotentialY = 96; PotentialY > 32; --PotentialY)
```

Comment:
Technically yes. If there was a 'ground' block there. So the only block which can cause an issue is the Dirt block as no other blocks should exist in a cave.

The problem with cave support again comes down to the Nether being a massive cave, so I had to support it to support Nether spawns.
If that is something that isn't wanted then I can add cave protection back in.

Personally a spawn point in the Nether is something I wanted on my server (thus this change :D) but I understand it might not be wanted for the project overall.